pandas/core/window/numba_.py

```diff
@@ -0,0 +1,112 @@
+import types
+from typing import Callable, Dict, Optional, Tuple
+
+import numpy as np
+
+from pandas.compat._optional import import_optional_dependency
+
+
+def generate_numba_apply_func(
+    args: Tuple,
+    kwargs: Dict,
+    func: Callable,
+    engine_kwargs: Optional[Dict],
+    function_cache: Dict,
+):
+    """
+    Generate a numba jitted apply function specified by values from engine_kwargs.
+
+    1. jit the user's function
+    2. Return a rolling apply function with the jitted function inline
+
+    Configurations specified in engine_kwargs apply to both the user's
+    function _AND_ the rolling apply function.
+
+    Parameters
+    ----------
+    args : tuple
+        *args to be passed into the function
+    kwargs : dict
+        **kwargs to be passed into the function
+    func : function
+        function to be applied to each window and will be JITed
+    engine_kwargs : dict
+        dictionary of arguments to be passed into numba.jit
+    function_cache : dict
+        dictionary of cached apply function to avoid re-compiling the apply loop
+
+    Returns
+    -------
+    Numba function
+    """
+    numba = import_optional_dependency("numba")
+
+    if engine_kwargs is None:
+        engine_kwargs = {}
+
+    nopython = engine_kwargs.get("nopython", True)
+    nogil = engine_kwargs.get("nogil", False)
+    parallel = engine_kwargs.get("parallel", False)
+
+    if kwargs and nopython:
+        raise ValueError(
+            "numba does not support kwargs with nopython=True: "
+            "https://github.com/numba/numba/issues/2916"
+        )
+
+    if parallel:
+        loop_range = numba.prange
+    else:
+        loop_range = range
+
+    # Return an already compiled version of roll_apply if available
+    if func in function_cache:
+        return function_cache[func]
+
+    def make_rolling_apply(func):
```

Comment:
can you move make_rolling_apply to module scope (out of this function)?

also don't you need to actually assign to the cache? (on a miss)

Reply:
The cache assignment happens here https://github.com/pandas-dev/pandas/pull/30151/files#diff-0de5c5d9abfcdd141e83701eaaec4358R541 (the function needs to run on some data first)

Reply:
then i wouldn’t even check the cache here ; that must be at the higher level

Reply:
Okay sure thing. I'll move it up then.